### PR TITLE
Merge 4.04 into trunk

### PR DESCRIPTION
**Pull request by gasche**
I had fun using the fairly convenient tool [git-imerge](https://github.com/mhagger/git-imerge) to try to merge the 4.04 branch into `trunk`. @damiendoligez, I know you generally take care of this after the release, but I am assuming you had been delayed somehow; is that PR acceptable to do this?

The commits in the list are the 4.04 commit. The real action is in the merge commit at the end, bb280404ed3da369bcd2b0c684f66be2960a0d25, but github does not show it in a convenient way at all! So I took the much nicer output of `git show` on this merge commit and added it to the PR as a new file [merge-commit.md](https://github.com/gasche/ocaml/blob/trunk-with-4.04/merge-commit.md) in commit 422f1097193713f796864280c55627d24b751447. Please look at this file to review the merge commit, and feel free to use annotations on the corresponding commit to indicate any part that looks odd to you.

### CODE REVIEW

**Comment by damiendoligez** [comment]
> @damiendoligez, I know you generally take care of this after the release, but I am assuming you had been delayed somehow; is that PR acceptable to do this?

Of course I'm happy if you take this burden off my shoulders.

**Comment by gasche** [comment]
Indeed you found a case where the automatic merge is not what we want -- at all.

The process I used to do the merge is somewhat naive: I just merged the branch `4.04` intro `trunk`. I have no experience with release-branch merges, maybe a better way to do it would be to first clone `4.04` into a branch from which we remove all release-process-specific things, keeping only the actual changes that went into 4.04, and then merge that. But then I'm not sure how to make git "see" that it is a merge of the 4.04 branch (maybe we don't need this at all).

(Do you also use [git imerge](https://github.com/mhagger/git-imerge) ?)

My proposal to answer your question of "How can we tell what else we're missing?" is to go over the commit log, find the "release process" commits, and check that their effect is gone from the post-merge branch or explicitly revert them. And then do a lot of testing in trunk, of course. Does that sound right?

**Comment by damiendoligez** [comment]
My process is:
- do the merge and resolve conflicts
- review the changes
- remove the (very few) branch-specific changes
- commit the merge

Sometimes, a change was done very differently in the branch and in the trunk. In this case, I just drop the branch changes after careful examination. In case of doubt I ask for a review from the original contributor.

Note that the camlp4 thing is branch-specific but not what I would call "release-process-specific" so I don't think it's possible to eliminate all such changes in advance.

But my question was: why does `git show` hide the camlp4 change and what else does it hide?

**Comment by gasche** [comment]
By default the `git show` output in [merge-commit.md](https://github.com/gasche/ocaml/blob/trunk-with-4.04/merge-commit.md) shows what git calls a "combined diff", which contains the conflict resolution information, that is those of the diffs of the two branches that were conflicting, merged with the changes I made manually to resolve the conflict. There are two `+`/`-` columns in the output: `+` is an addition coming from the merged branch (4.04), `+` is an addition coming from the trunk, and `++` is an addition coming from my merge resolution.

The camlp4 change from 4.04 does not conflict with anything in trunk, so it does not show in the combined diff.

To see the full diff, you can use `git show -m` (`-m` for "merge"; `--cc` for "combined" (default)). This is what github shows when looking at the commit on their interface. The full merge diff shows over 32k lines (with 21k changed lines), compared with 5k lines for the combined diff (3k changed lines), so I wouldn't want to review all of them.

**Comment by damiendoligez** [comment]
> The full merge diff shows over 32k lines (with 21k changed lines), compared with 5k lines for the combined diff (3k changed lines), so I wouldn't want to review all of them.

And yet, it needs to be done.

**Comment by alainfrisch** [comment]
The (currently 32k lines) diff would have been much smaller if most commits on 4.04 were cherry-picked to trunk "on the fly", by each developer, right?  Unless we manage to bring down the time between branching and release very significantly, I think we should strongly encourage / mandate that each developer committing to a release branch also port the fix to trunk at the same time.  The side effect will be to make it more painful to commit to a release branch (and I would argue this is a good side effect).

This will not solve the problem of fixes that come with a very different version between the release branch and trunk.  Since git has no way to be informed that a specific commit from a branch has been "manually" integrated to trunk (contrary to svn), one possibility would be to merge very regularly the release branch into trunk, perhaps even after any commit to the release branch.

Globally, I think one should aim at:
- keeping trunk up to date (having a bug fixes on a release branch only and not trunk for several weeks is not good and it happened);
- putting the burden of adapting fixes for both a release branch and trunk on the developer who submitted the fix, not on the release manager;
- making it more painful to commit to a release branch than to trunk only, to give everyone incentives to "play nice" with our release cycle.

@damiendoligez @gasche Do you agree?

**Comment by gasche** [comment]
I am not the release manager (fortunately) and I would let Damien make decision on the workflow that works best for him. That said, I agree that merging release branches into trunk regularly would be a good way to smooth the process.

In fact I may be able to do this after the fact by going back to my incremental merge, and splitting it in chunks as if several merges had been done during the development of the 4.04 branch. I'm not sure I can go back to this (it was my first experience with [git imerge](https://github.com/mhagger/git-imerge) and I did a few missteps), but if it works it could be a way to review the merges separately and thus more conveniently. One small merge a week avoids reviewer panic!

I'll try to give this a try in the next few days (not today) and report back.

**Comment by damiendoligez** [comment]
> The (currently 32k lines) diff would have been much smaller if most commits on 4.04 were cherry-picked to trunk "on the fly", by each developer, right?

Let me rephrase that: the diff would have been much larger if most commits had _not_ been cherry-picked to trunk...

Maybe it would be a good idea to merge the branch into trunk more often, but in general the merge is not such a big issue. This time it's a bit bigger than usual.

**Comment by damiendoligez** [comment]
I have reviewed the diff and it looks OK except for the camlp4 line.

**Comment by gasche** [comment]
Wow. Thank you! I think I need to rebase the merge anyway (not sure how to do that), I'll also remove this line, and update the PR.

**Comment by damiendoligez** [comment]
> I need to rebase the merge anyway (not sure how to do that),

Ouch. First prepare some painkillers, then search for "git rerere" in your favorite search engine. Unfortunately, it only helps after you activate it, so for this time you'll have to redo all the conflict resolution :-(

**Comment by mshinwell** [comment]
@gasche Has there been any more progress on this one?

**Comment by dra27** [comment]
I started doing a normal rebase last night, but I won't be able to go back to it until tomorrow morning (cherry-picking testsuite fixes was proving a nuisance!).

**Comment by damiendoligez** [comment]
I'm now trying to do the merge old-fashioned. Don't work on it until I report here.

**Comment by dra27** [comment]
#962 has been merged.
